homeassistant/components/openweathermap/forecast_update_coordinator.py

```diff
@@ -0,0 +1,137 @@
+from pyowm.exceptions.api_response_error import UnauthorizedError
+
+from homeassistant.components.weather import (
+    ATTR_FORECAST_CONDITION,
+    ATTR_FORECAST_PRECIPITATION,
+    ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME,
+    ATTR_FORECAST_WIND_BEARING,
+    ATTR_FORECAST_WIND_SPEED,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    ATTR_API_FORECAST,
+    ATTR_API_THIS_DAY_FORECAST,
+    CONDITION_CLASSES,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+FORECAST_UPDATE_INTERVAL = timedelta(minutes=30)
+
+
+class ForecastUpdateCoordinator(DataUpdateCoordinator):
+    """Forecast data update coordinator."""
+
+    def __init__(self, owm, latitude, longitude, forecast_mode, hass):
+        """Initialize coordinator."""
+        self._owm_client = owm
+        self._forecast_mode = forecast_mode
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecast_limit = 15
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=FORECAST_UPDATE_INTERVAL
+        )
+
+    async def _async_update_data(self):
+        data = {}
+        with async_timeout.timeout(20):
+            try:
+                forecast_response = await self._update_forecast()
+                data = self._convert_forecast_response(forecast_response)
+            except (APICallError, UnauthorizedError) as error:
+                raise UpdateFailed(error)
+
+        return data
+
+    async def _update_forecast(self):
+        if self._forecast_mode == "daily":
+            forecast_response = await self.hass.async_add_executor_job(
+                self._owm_client.daily_forecast_at_coords,
+                self._latitude,
+                self._longitude,
+                self._forecast_limit,
+            )
+        else:
+            forecast_response = await self.hass.async_add_executor_job(
+                self._owm_client.three_hours_forecast_at_coords,
+                self._latitude,
+                self._longitude,
+            )
+        return forecast_response.get_forecast()
+
+    def _convert_forecast_response(self, forecast_response):
+        values = self._get_weathers(forecast_response)
+
+        forecast_entries = self._convert_forecast_entries(values)
+
+        return {
+            ATTR_API_FORECAST: forecast_entries,
+            ATTR_API_THIS_DAY_FORECAST: forecast_entries[0],
+        }
+
+    def _convert_forecast_entries(self, values):
+        if self._forecast_mode == "daily":
+            return list(map(self._convert_daily_forecast, values))
+        return list(map(self._convert_forecast, values))
+
+    def _get_weathers(self, forecast_response):
+        if self._forecast_mode == "freedaily":
+            return forecast_response.get_weathers()[::8]
+        return forecast_response.get_weathers()
+
+    def _convert_daily_forecast(self, entry):
```

Comment:
Please make the naming consistent.  Here its called `entry`, and in  `_convert_forecast_entries` its called `values`